### ISSUE TITLE
Bayesian R square for greta model

### ISSUE THREAD

**Issue by lionel68** [greta-dev/greta#345]
Inspired by the work from Gelman and others on [Bayesian R-square](https://stat.columbia.edu/~gelman/research/published/bayes_R2_v3.pdf), I was wondering if it would make sense to include such a function in greta or in an associated packages to run model checks on greta model fits.

In essence one needs two things:

- the vector of response values (y)
- the posterior draws from the linear predictor (which we could get from _calculate_) on the response scale (let's call this ypred)

Then in R code we have:

```r
# posterior residuals
e <- -1 * sweep(as.matrix(ypred), 2, y)
# variance in linear predictors
var_ypred <- apply(as.matrix(ypred), 1, var)
# variance in residuals
var_e <- apply(e, 1, var)
# R2 values
post_r2 <- var_ypred / (var_ypred + var_e)
```

The trick is basically to get the posterior draws from the linear predictors.

Any ideas?

**Comment by goldingn** [comment]
Cool. I've not looked into this before, but it looks useful.

I'm not sure whether it would fit in the main Greta package, or could automatically be added to a summary of the draws object, since it doesn't appear to apply to all models. But it would be nice to provide in an extension package of valuation and model checking metrics.

My interpretation it's only applicable to glm-esque models. Or, more specifically, models that have an observed (I.e. data following a distribution) vector variable with each element following a univariate distribution. There would also have to be an expression for the expected value of the data under the distribution. Ie. this would be straightforward for a Bayesian linear model or, say, Poisson regression, but wouldn't exist for a simple model without observed data like this:
```r
sd <- normal(0, 1, truncation = c(0, Inf))
mu <- normal(0, 1)
x <- normal(mu, sd)
```
and would be tricky for an observed variable where the expectation isn't a parameter of the distribution:
```r
distribution(y) <- beta(a, b)
```

Maybe an interface where the user could pass in greta arrays for the data and prediction, and some posterior samples would be best? E.g.:
```r
distribution(y) <- beta(a, b)
m <- model(a, b)
draws <- mcmc(m)

pred <- a / (a + b)
r2(y, pred, draws)
```

**Comment by lionel68** [comment]
Yes, such R2 only make sense for GLM(M)s but I find it a useful metric.

Your idea of an interface where the user passes greta arrays for the data and the prediction plus the posterior draws sounds like it could do the job.

I guess I'll start getting my hands dirty and would put this into an extension package. Another idea would be to allow users to test their GLMesque models via [DHARMa](https://github.com/florianhartig/DHARMa/wiki/Adding-new-R-packages-to-DHARMA), given your recent implementation of the simulate function this should be doable.

**Comment by lionel68** [comment]
All right, a first try is now here: https://github.com/lionel68/greta.checks/blob/master/R/bayes_R2.r

**Comment by goldingn** [comment]
Nice! `greta.checks` a perfect name for this package
